### DISCUSSION TITLE
bash-language-server: avoid recursive workspace scanning

### DISCUSSION THREAD

**Discussion by akurth** [joaotavora/eglot#955] [Q&A]
Hello,

if I open an sh file in $HOME (say ~/.zshrc ), bash-language-server scans the entire directory for sh files recursively, which takes very long and uses a lot of CPU unnecessarily.

I guess it’s because workspaceFolders URI has been set to my home directory. Is there any way to set it to nil, or another way of disabling workspace scanning for this particular server?

Would appreciate any suggestions.

**Comment by nemethf** [comment]
One possible approach would be to create a project.el backend for which `project-files` doesn't traverse into sub-directories or returns files that the user explicitly defined as part of the project.

**Reply by joaotavora** [reply]
Sounds smart.  

Another option, if I am right to assume you have your whole $HOME directory under Git version control, which `project.el` deals with by default is to use `.gitignore` or something so that `project.el` is informed of this.

Another option yet might be to use symbolic links.  My `~/.zshrc` points into something like `~/foo/library-of-sh-code/zee-shell-arrcee.zsh`.  I've yet to discover how `M-x eglot` with being invoked on symbolic links though.  I don't think it'll send the referent directory to the server.  But at least Emacs tells you that you're in a symlink and makes it easy to open the referent.  Then you could `M-x eglot` there.

**Comment by akurth** [comment]
My home directory is not under version control. And my .zshrc is a plain file, since I’m using chezmoi to manage my config files.

I’m not sure whether project.el is involved here at all (isn’t ~ the default workspace root?), but I’ll have a look at the project backend suggestion.

**Reply by joaotavora** [reply]
`project.el` is definitely involved here: it's whom Eglot asks to say where the "project root" of the file you invoked `M-x eglot` in lives.  

But I see what you mean: reading your descirption more closely, it is the LSP `bash-language-server` who is doing copious amounts of scanning (indeed `project.el`/Eglot doesn't do much of that).  Then I suppose the problem is in the LSP server's behaviour.  If you're opening a file in `foo/bar.bash` which other directory _but_ `foo` can you supply to the LSP server?

**Reply by nemethf** [reply]
> But I see what you mean: reading your descirption more closely, it is the LSP `bash-language-server` who is doing copious amounts of scanning 

Isn't the server's behavior the result of Eglot's way of handling dynamic registration of workspace/didChangeWatchedFiles? I mean:

https://github.com/joaotavora/eglot/blob/d5e335f299216bdf3fe4b0d71fb6e24596a35a6a/eglot.el#L3063-L3066

---

I don't know what chezmoi is, but according to the docs, it seems to use a git repo under the hood  in ~/.local/share/chezmoi.

**Reply by joaotavora** [reply]
> Isn't the server's behavior the result of Eglot's way of handling dynamic registration of workspace/didChangeWatchedFiles? I mean:

That doesn't count as "scanning" for me, rather, "watching".  And Eglot, the LSP client, only sets watchers when the server requests it to.  But without more information, it's indeed unknown what is happening.  I read:

> > bash-language-server scans the entire directory for sh files recursively,

.. as meaning that the LSP server is doing lots of scanning and that @akurth is looking for a (LSP) way to convince it to do less...

**Reply by akurth** [reply]
Please see the events log for ~/.zshrc here: https://nc.andreas-kurth.de/index.php/s/A6CW92TyxSzeZN7
(I deleted hundreds of similar "Analyzing file" lines and the contents of my .zshrc.)

If I open ~/bin/foo.sh, the first client request is:
`(:jsonrpc "2.0" :id 1 :method "initialize" :params (:processId 17940 :rootPath "/home/ak/bin/" :rootUri "file:///home/ak/bin"`, and only the SH files within ~/bin are scanned.

**Reply by akurth** [reply]
This sounds related and seems to be what I’d like to have: https://github.com/muniter/neovim/commit/ab86306ca0bea8f27f97f1cb573a78cb10c05ff1

**Reply by joaotavora** [reply]
Yes. At least this confirms it's the LSP server doing the work.  You might have some some luck turning off the verbose logging either on the server on on Eglot's side.  See `eglot-events-buffer-size`.  Though if it's "hundreds" of lines and not "thousands+", I don't think it'll make much of a difference.
